functions.lua

```diff
@@ -241,6 +241,72 @@ function PacifistMod.remove_misc()
         assert(entry[1] and entry[2])
         data_raw.remove(entry[1], entry[2])
     end
+
+    -- hide explosion entities revealed by removing/hiding other things
+    data_raw.hide("explosion", "atomic-nuke-shockwave")
+    data_raw.hide("explosion", "wall-damaged-explosion")
+end
+
+function PacifistMod.enumerate_used_names()
+    local all_names = {}
+    for _, list in pairs(data.raw) do
+        for _, entity in pairs(list) do
+            all_names[entity.name] = true
+        end
+    end
+
+    local referenced_names = {}
+
+    local function enumerate_all_strings(x)
+        if not x then
+            return
+        elseif type(x) == "string" then
+            if all_names[x] then
+                referenced_names[x] = true
+            end
+        elseif type(x) == "table" then
+            for name, el in pairs(x) do
+                if not (name == "type") then
+                    enumerate_all_strings(name)
+                    enumerate_all_strings(el)
+                end
+            end
+        end
+    end
+
+    for _, list in pairs(data.raw) do
+        for _, entry in pairs(list) do
+            for name, value in pairs(entry) do
+                if not (name == "name" or name == "type") then
+                    enumerate_all_strings(value)
+                end
+            end
+        end
+    end
+
+    return referenced_names
+end
+
+function PacifistMod.remove_orphaned_entities(was_used)
+    repeat
+        local changed = false
+        local is_used = PacifistMod.enumerate_used_names()
```

Comment:
Hmm... I haven't tried it, I can test with SE/Seablock/other overhauls. But I think I can rework the algorithm to only do the full read of `data.raw` once and record which things reference which names instead of just whether a name is referenced. If we also had a list of everything Pacifist removes (which shouldn't be too hard to maintain...), then we can avoid ever doing any more iterations of even the list of names in `data.raw` (at that point, it's basically a reference counting garbage collector). Which is only moderately more complicated and should be notably faster.

Reply:
SE isn't compatible, but Seablock should be, as well as Exotic Industries - those are so far the largest mod packs I'd like to support

Reply:
FYI, SeaBlock fails to load. I added it as an optional dependency because it was erroring on something added after Pacifist ran, but now it says

> Technology military: Non-contiguous levels: 1, followed by 3

and I did not investigate further.

Reply:
I have a draft of the new algorithm in a branch https://github.com/dperelman/Pacifist/tree/feature/remove-orphaned-entities-opt . I'll merge that into this PR after I do a bit more clean up and comment writing.

Here's my performance measurements. Numbers are the timestamp on the log line that says "Factorio initialised", which I believe is seconds since Factorio started. I ran each configuration a few times and took the last one to avoid caching effects (the numbers usually dropped slightly after the first run). The left column is with just the base mod, the right column is with "Exotic Industries" loaded (i.e. just the default of whatever that mod depends on, no optional dependencies). Each row is a different configuration of Pacifist. The third row is this PR as it currently is, the final row is the branch linked at the start of this comment as it currently is.

Surprisingly, Pacifist makes the game load significantly *faster*, I guess because it's removing things, so later phases of the loading process have less work to do?
| mod           | base    | EI     |
|:--------------|--------:|-------:|
| no Pacifist   |  13.181 | 19.802 |
| `main`        |  11.404 | 16.909 |
| naive orphans |  10.897 | 17.563 |
| opt orphans   |   9.617 | 14.936 |